### ISSUE TITLE
Computing condition first in where

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#2579]
Have recently encountered the need for computing the condition first in Dask Array' `where`. Normally one can use `where` pretty effectively when trying to select from two arrays. Unfortunately this requires evaluating both arrays that go to `where` (in addition to the `condition`). This is fine when both arrays can be evaluated reasonably, but when they can't this is problem.

The (pedestrian) example below shows this behavior where we attempt to side step an exception by using `where`. However we still get the exception even though the condition should forbid it as all arrays are evaluated (calling `compute` on the `condition` in `where` verifies this). One could argue that we should just catch the exception here and handle it. Though this breaks down quickly as we may not have access to the call to `compute`. We can try to remedy this in other ways by using things like `delayed`. However this doesn't seem to perform well.

An ideal fix would be to compute the condition first and from that determine, which, if any, of the other arrays need to be computed. Thus only computing them as needed. It should also improve performance of `where` in other cases as well. Also it should still respect chunks in its implementation.

Example:

<details>

```python
In [1]: import numpy as np

In [2]: import dask.array as da

In [3]: a = np.array([0.0, -0.5], dtype=np.float64)

In [4]: d = da.asarray(a)

In [5]: m = (d > 0)

In [6]: r = da.where(m.any(), d[m].min(), 1)

In [7]: r.compute()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-636a08ef9c75> in <module>()
----> 1 r.compute()

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/base.pyc in compute(self, **kwargs)
     96             Extra keywords to forward to the scheduler ``get`` function.
     97         """
---> 98         (result,) = compute(self, traverse=False, **kwargs)
     99         return result
    100 

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/base.pyc in compute(*args, **kwargs)
    203     dsk = collections_to_dsk(variables, optimize_graph, **kwargs)
    204     keys = [var._keys() for var in variables]
--> 205     results = get(dsk, keys, **kwargs)
    206 
    207     results_iter = iter(results)

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/threaded.pyc in get(dsk, result, cache, num_workers, **kwargs)
     73     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     74                         cache=cache, get_id=_thread_get_id,
---> 75                         pack_exception=pack_exception, **kwargs)
     76 
     77     # Cleanup pools associated to dead threads

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/local.pyc in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    519                         _execute_task(task, data)  # Re-execute locally
    520                     else:
--> 521                         raise_exception(exc, tb)
    522                 res, worker_id = loads(res_info)
    523                 state['cache'][key] = res

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/local.pyc in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    288     try:
    289         task, data = loads(task_info)
--> 290         result = _execute_task(task, data)
    291         id = get_id()
    292         result = dumps((result, id))

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/local.pyc in _execute_task(arg, cache, dsk)
    268     elif istask(arg):
    269         func, args = arg[0], arg[1:]
--> 270         args2 = [_execute_task(a, cache) for a in args]
    271         return func(*args2)
    272     elif not ishashable(arg):

/zopt/conda2/envs/test/lib/python2.7/site-packages/dask/local.pyc in _execute_task(arg, cache, dsk)
    269         func, args = arg[0], arg[1:]
    270         args2 = [_execute_task(a, cache) for a in args]
--> 271         return func(*args2)
    272     elif not ishashable(arg):
    273         return arg

/zopt/conda2/envs/test/lib/python2.7/site-packages/numpy/core/fromnumeric.pyc in amin(a, axis, out, keepdims)
   2370 
   2371     return _methods._amin(a, axis=axis,
-> 2372                           out=out, **kwargs)
   2373 
   2374 

/zopt/conda2/envs/test/lib/python2.7/site-packages/numpy/core/_methods.pyc in _amin(a, axis, out, keepdims)
     27 
     28 def _amin(a, axis=None, out=None, keepdims=False):
---> 29     return umr_minimum(a, axis, None, out, keepdims)
     30 
     31 def _sum(a, axis=None, dtype=None, out=None, keepdims=False):

ValueError: zero-size array to reduction operation minimum which has no identity
```

</details>

<br>
Environment:
<br>
<br>

<details>

```yaml
name: test
channels:
- conda-forge
- defaults
dependencies:
- appnope=0.1.0=py27_0
- asn1crypto=0.22.0=py27_0
- backports.shutil_get_terminal_size=1.0.0=py27_1
- backports_abc=0.5=py27_0
- bkcharts=0.2=py27_0
- blas=1.1=openblas
- bokeh=0.12.6=py27_0
- ca-certificates=2017.4.17=0
- certifi=2017.4.17=py27_0
- cffi=1.10.0=py27_0
- chardet=3.0.2=py27_1
- click=6.7=py27_0
- cloudpickle=0.3.1=py27_0
- cryptography=1.9=py27_0
- cycler=0.10.0=py27_0
- dask=0.15.1=py27_0
- decorator=4.1.2=py27_0
- distributed=1.18.0=py27_0
- enum34=1.1.6=py27_1
- freetype=2.7=1
- functools32=3.2.3.2=py27_1
- futures=3.0.5=py27_0
- heapdict=1.0.0=py27_0
- idna=2.5=py27_0
- ipaddress=1.0.18=py27_0
- ipython=5.4.1=py27_0
- ipython_genutils=0.2.0=py27_0
- jinja2=2.9.5=py27_0
- libffi=3.2.1=3
- libgfortran=3.0.0=0
- libpng=1.6.28=0
- locket=0.2.0=py27_1
- markupsafe=1.0=py27_0
- matplotlib=2.0.2=np113py27_0
- msgpack-python=0.4.8=py27_0
- ncurses=5.9=10
- numpy=1.13.1=py27_blas_openblas_201
- openblas=0.2.20=1
- openssl=1.0.2l=0
- pandas=0.20.3=py27_1
- partd=0.3.8=py27_0
- pathlib2=2.3.0=py27_0
- pexpect=4.2.1=py27_0
- pickleshare=0.7.3=py27_0
- pip=9.0.1=py27_0
- prompt_toolkit=1.0.14=py27_0
- psutil=5.2.1=py27_0
- ptyprocess=0.5.2=py27_0
- pycparser=2.18=py27_0
- pygments=2.2.0=py27_0
- pyopenssl=16.2.0=py27_0
- pyparsing=2.2.0=py27_0
- pysocks=1.6.7=py27_0
- python=2.7.13=1
- python-dateutil=2.6.1=py27_0
- pytz=2017.2=py27_0
- pyyaml=3.12=py27_1
- readline=6.2=0
- requests=2.18.2=py27_0
- scandir=1.5=py27_1
- setuptools=36.2.2=py27_0
- simplegeneric=0.8.1=py27_0
- singledispatch=3.4.0.3=py27_0
- six=1.10.0=py27_1
- sortedcontainers=1.5.3=py27_0
- sqlite=3.13.0=1
- ssl_match_hostname=3.5.0.1=py27_1
- subprocess32=3.2.7=py27_0
- tblib=1.3.2=py27_0
- tk=8.5.19=1
- toolz=0.8.2=py27_0
- tornado=4.5.1=py27_0
- traitlets=4.3.2=py27_0
- urllib3=1.21.1=py27_1
- wcwidth=0.1.7=py27_0
- wheel=0.29.0=py27_0
- yaml=0.1.6=0
- zict=0.1.2=py27_0
- zlib=1.2.11=0
- pip:
  - backports-abc==0.5
  - backports.shutil-get-terminal-size==1.0.0
  - backports.ssl-match-hostname==3.5.0.1
  - ipython-genutils==0.2.0
  - prompt-toolkit==1.0.14
```

</details>

**Comment by jcrist** [comment]
That's not the semantics of `np.where`. In your example above you'd get the same error if you were using numpy:

```python
In [1]: import numpy as np

In [2]: a = np.array([0.0, -0.5], dtype=np.float64)

In [3]: m = (a > 0)

In [4]: np.where(m.any(), a[m].min(), 1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-8256f34227db> in <module>()
----> 1 np.where(m.any(), a[m].min(), 1)

~/anaconda/envs/dask/lib/python3.5/site-packages/numpy/core/_methods.py in _amin(a, axis, out, keepdims)
     27
     28 def _amin(a, axis=None, out=None, keepdims=False):
---> 29     return umr_minimum(a, axis, None, out, keepdims)
     30
     31 def _sum(a, axis=None, dtype=None, out=None, keepdims=False):

ValueError: zero-size array to reduction operation minimum which has no identity
```

> It should also improve performance of where in other cases as well.

I don't think this is universally true. In many cases dask arrays represent long computations. In your case above, the condition and the value share the same graph, which would then be evaluated twice - once for computing the initial condition, and once for computing the resulting value. Further, the condition may not fully fit in memory, losing all out-of-core benefits. As such, the proposed solution is a no-go.

---

I think what you're looking for is conditional application based on value. Something like:

```python
@dask.delayed
def conditional_apply(x, cond, true, false):
    if cond(x):
        return true(x)
    return false(x)
```

in the scalar case, or something similar + `da.map_blocks` in the array case. I can't offer a concrete suggestion I guess without seeing what the problem you're trying to solve is (assuming the example above was not exactly it).

**Comment by jakirkham** [comment]
> That's not the semantics of np.where. In your example above you'd get the same error if you were using numpy...

I'm not sure this is entirely fair. NumPy is not lazy. Dask is. So would expect Dask could be a bit smarter about what it evaluates (or skips). In any event, had considered making a request for a lazy `if` instead of asking about `where`. So the same question can be recast in that context.

>In your case above, the condition and the value share the same graph, which would then be evaluated twice - once for computing the initial condition, and once for computing the resulting value.

Admittedly this example is a bit contrived. So maybe it isn't worth thinking about that much.

> Further, the condition may not fully fit in memory, losing all out-of-core benefits. As such, the proposed solution is a no-go.

As noted above, am proposing that this respect chunking. IOW it would evaluate a condition chunk and then determine whether the corresponding `x` or `y` chunk should be evaluated before doing so. So I fail to see why this is true.

> I think what you're looking for is conditional application based on value.

That's what I ended up doing for now. Though I'm a little concerned about the performance of this solution. Hence raising this issue.

> I can't offer a concrete suggestion I guess without seeing what the problem you're trying to solve is (assuming the example above was not exactly it).

Sure. If you have some better ideas for my use case, would definitely appreciate the feedback. Please see [lines 142-145]( https://github.com/dask-image/dask-ndmeasure/pull/30/files#diff-1799b6e9926daa3f6fda9bee52080e9eR142 ). Thanks.

**Comment by jcrist** [comment]
> Sure. If you have some better ideas for my use case, would definitely appreciate the feedback. Please see lines 142-145. Thanks.

I'm not familiar with that scipy function, but it looks like you're doing a loop over all indices in a dask array, and computing a function on them? If so, that'd create many many tasks, each of which take only a short amount of time.

Instead I'd write this blockwise using `da.atop`. Something like (untested):

```python
def conditional_apply(cond, x, default=None, func=None):
    # Apply `func` only to those elements where `cond`
    # is true, fill in default everywhere else
    assert cond.shape == x.shape
    out = np.full(x.shape, default)
    out[cond] = func(x[cond])
    return out

inds = tuple(range(x.ndim))
res = da.atop(conditional_apply, inds, cond, inds, x, inds, default=default, func=func)
# then reshape res to whatever the output shape is supposed to be (if needed?)
```

**Comment by jakirkham** [comment]
> ...it looks like you're doing a loop over all indices in a dask array, and computing a function on them?

Sorry, I should have explained this first.

This is a more detailed explanation if the TL;DR doesn't make sense. The pseudocode in the SciPy docs `[func(input[labels == i]) for i in index]` is somewhat helpful in getting context. Here `input` is some image, `labels` are coded groupings (i.e. objects) within that image, and `index` are which groupings to compute over. This [blog entry]( http://emmanuelle.github.io/a-tutorial-on-segmentation.html ) provides a non-trivial segmentation example (no need to read it all). The first two images could be seen as `input` and `labels` in that order. So the user may want to compute some property (e.g. max intensity) of the big yellow object in the middle (from the top right image in the blog entry). To provide more context this is often used to subselect from the objects in the `label` image (e.g. types of cells).

TL;DR `labeled_comprehension` computes a user specified reduction (e.g. max intensity) on labels (i.e. user selected regions) of an image.

Hope that helps and is the right level of detail. If that doesn't make sense, please feel free to ask question.

**Comment by mrocklin** [comment]
> An ideal fix would be to compute the condition first and from that determine, which, if any, of the other arrays need to be computed. Thus only computing them as needed. It should also improve performance of where in other cases as well. Also it should still respect chunks in its implementation.

I agree that this would be helpful.  Unfortunately the current schedulers do not support this kind of conditional computation.  I agree that that would be a useful feature to the schedulers.  I personally do not have any plans to implement this near-term.  It would be a non-trivial change for the single-machine scheduler.

**Comment by mrocklin** [comment]
It sounds like this is out of scope today.  This conversation has gone stale. Closing.

**Comment by jakirkham** [comment]
Think we made a few low-level optimizations to `da.where` like handling a concrete `bool` conditional specifically, which is useful. Some high-level form of branching would still be useful, but probably belongs in a different issue entirely.
